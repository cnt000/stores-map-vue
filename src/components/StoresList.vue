<template>
  <div class="stores">
    <select @change="selectCountryId" v-model="countryId" class="stores_countries">
      <option value="0">All</option>
      <option
        v-for="country in countries"
        :key="country.term_id"
        :value="country.term_id"
      >{{ country.name }}</option>
    </select>
    <div>
      <input class="stores_search" type="text" placeholder="filter" v-model="keyword">
      <span v-if="pending">... LOADING ...</span>
      <span v-if="error">... ERROR!!! ...</span>
      <ul class="stores_list">
        <li
          v-if="filteredStores.length === 0"
          class="stores_list_store stores_list_store--empty"
        >Non ci sono risultati</li>
        <li
          v-else
          v-for="store in filteredStores"
          :data-storeid="store.ID"
          :key="store.ID"
          v-on:click="selectStore(store.ID)"
          class="stores_list_store"
        >
          <span class="stores_list_store_name">{{ store.post_title }}</span>
          <span>{{ store.custom["wpcf-yoox-store-address"][0] }}</span>
          <span>({{ store.lat }} - {{ store.lng }})</span>
        </li>
      </ul>
    </div>
  </div>
</template>

<script>
import { mapState } from "vuex";
import * as R from "ramda";

export default {
  data() {
    return {
      keyword: "",
      countryId: 0
    };
  },
  computed: {
    ...mapState({
      stores: state => state.stores.all,
      countries: state => state.stores.countries,
      CountryId: state => state.stores.selectedCountryId,
      pending: state => state.stores.pending,
      error: state => state.stores.error
    }),
    filteredStores() {
      const keywordInLowerCase = R.toLower(this.keyword);
      const hasKeywordInTitle = R.includes(keywordInLowerCase);
      const storeName = R.prop("post_title");
      const filterByKeyword = R.filter(
        R.compose(
          hasKeywordInTitle,
          R.toLower,
          storeName
        )
      );
      const countryId = this.CountryId;
      const hasCountryId = R.any(R.propEq("term_id", countryId));
      const filterByCountryId = R.filter(
        R.compose(
          hasCountryId,
          R.prop("terms")
        )
      );
      const countryIdGtZero = () => R.gt(this.CountryId, 0);
      const filterByCountryIdAndKeyword = R.compose(
        filterByKeyword,
        R.when(countryIdGtZero, filterByCountryId)
      );
      return filterByCountryIdAndKeyword(this.stores);
    }
  },
  methods: {
    selectStore(clickedId) {
      this.$store.dispatch({
        type: "stores/selectStore",
        id: clickedId
      });
    },
    selectCountryId() {
      this.$store.dispatch({
        type: "stores/selectCountryId",
        id: +this.countryId
      });
    }
  },
  created() {
    this.$store.dispatch("stores/getCountries");
  }
};
</script>

<style lang="scss">
.stores {
  &_search {
    margin-top: 2px;
    box-sizing: border-box;
    border: 1px solid black;
    width: 100%;
    height: 36px;
    font-size: 18px;
    border-bottom: 0;
    text-indent: 8px;
    &::placeholder {
      font-size: 14px;
    }
  }
  &_countries {
    width: 100%;
    height: 60px;
  }
  &_list {
    border-top: 1px solid black;
    width: 100%;
    list-style: none;
    margin: 0;
    padding: 0;
    &_store {
      display: flex;
      flex-direction: column;
      border: 1px solid black;
      border-top: none;
      padding: 4px;
      cursor: pointer;
      &--empty {
        background-color: lightsalmon;
      }
      &:hover {
        background-color: lightgreen;
      }
      &:active {
        background-color: lightcoral;
      }
      &_name {
        font-weight: bold;
      }
    }
  }
}
</style>
