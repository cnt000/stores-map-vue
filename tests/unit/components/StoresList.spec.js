import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import StoresList from "@/components/StoresList.vue";
import storesExamples from "../stores-examples.json";
import countriesExamples from "../countries-examples.json";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("StoresList.vue", () => {
  let actions;
  let store;
  let testState;

  beforeEach(() => {
    actions = {
      "stores/selectStore": jest.fn(),
      "stores/getCountries": jest.fn(),
      "stores/selectCountryId": jest.fn()
    };

    testState = {
      all: storesExamples,
      pending: false,
      error: false,
      selectedStoreId: 0,
      countries: countriesExamples,
      selectedCountryId: 0
    };

    store = new Vuex.Store({
      modules: {
        stores: {
          state: testState,
          actions
        }
      }
    });
  });

  it("renders list item ABU DHABI - SOWWAH - I element", () => {
    const msg = "ABU DHABI - SOWWAH";
    const wrapper = shallowMount(StoresList, {
      store,
      localVue
    });
    expect(wrapper.find("li:first-child > span").text()).toMatch(msg);
  });

  it("renders list item BAKU - PORT BAKU - II element", () => {
    const msg = "BAKU - PORT BAKU";
    const wrapper = shallowMount(StoresList, {
      store,
      localVue
    });
    expect(wrapper.find("li:nth-child(2) > span").text()).toMatch(msg);
  });

  it("renders countries select with All", () => {
    const msg = "All";
    const wrapper = shallowMount(StoresList, {
      store,
      localVue
    });
    expect(wrapper.find("select option").text()).toMatch(msg);
  });

  it("renders countries select with China second option", () => {
    const msg = "Asia"; // not important if country or continents in this test
    const wrapper = shallowMount(StoresList, {
      store,
      localVue
    });
    expect(wrapper.find("select option:nth-child(2)").text()).toMatch(msg);
  });

  it("renders one store searching for BAKU - PORT BAKU", () => {
    const msg = "BAKU - PORT BAKU";
    const wrapper = shallowMount(StoresList, {
      store,
      localVue
    });
    wrapper.find(".stores_search").element.value = msg;
    wrapper.find(".stores_search").trigger("input");
    expect(wrapper.find("li:first-child > span").text()).toMatch(msg);
  });

  it("renders any store searching for AAA", () => {
    const msg = "";
    const wrapper = shallowMount(StoresList, {
      store,
      localVue
    });
    wrapper.find(".stores_search").element.value = msg;
    wrapper.find(".stores_search").trigger("input");
    expect(wrapper.find("li:first-child > span").text()).toMatch(msg);
  });

  it('calls store action "selectStore" when list item is clicked', () => {
    const wrapper = shallowMount(StoresList, { store, localVue });
    const listItem = wrapper.find("li:first-child > span");
    listItem.trigger("click");
    expect(actions["stores/selectStore"]).toHaveBeenCalled();
  });

  it('calls store action "selectCountryId" when list item is clicked', () => {
    const wrapper = shallowMount(StoresList, { store, localVue });
    const selectCountry = wrapper.find("select");
    selectCountry.trigger("change", 0);
    expect(actions["stores/selectCountryId"]).toHaveBeenCalled();
  });

  it('calls store action "getCountries" when is created', () => {
    shallowMount(StoresList, { store, localVue });
    expect(actions["stores/getCountries"]).toHaveBeenCalled();
  });
  it("matches snapshot", () => {
    const wrapper = shallowMount(StoresList, { store, localVue });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
