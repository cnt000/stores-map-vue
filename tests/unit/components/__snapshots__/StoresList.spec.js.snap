// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoresList.vue matches snapshot 1`] = `
<div class="stores"><select class="stores_countries">
    <option value="0">All</option>
    <option value="164">Asia</option>
    <option value="165">China</option>
    <option value="168">Japan</option>
    <option value="194">Saudi Arabia</option>
  </select>
  <div><input type="text" placeholder="filter" class="stores_search">
    <!---->
    <!---->
    <ul class="stores_list">
      <li data-storeid="1742" class="stores_list_store"><span class="stores_list_store_name">ABU DHABI - SOWWAH</span> <span>The Galleria, Sowwah Square</span> <span>(24.50138013631387 - 54.38876131600796)</span></li>
      <li data-storeid="1744" class="stores_list_store"><span class="stores_list_store_name">BAKU - PORT BAKU</span> <span>Port Baku Mall, Neftchiler ave. 153</span> <span>(40.3751057 - 49.8636196)</span></li>
      <li data-storeid="1721" class="stores_list_store"><span class="stores_list_store_name">BEIJING - SANLITUN</span> <span>Sanlitun Village, Village North Shop N2 - 12/22 11 Sanlitun Road, Chaoyang District 100027</span> <span>(39.938991808861964 - 116.45540810317993)</span></li>
    </ul>
  </div>
</div>
`;
